amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java

```diff
@@ -224,12 +244,34 @@ public void cancel() {
     }
 
     @Override
-    public void notify(RecordsDeliveryAck ack) {
-        eventDeliveryLock.release();
+    public synchronized void notify(RecordsDeliveryAck recordsDeliveryAck) {
+        final RecordsRetrieved recordsToCheck = peekNextResult();
+        // Verify if the ack matches the head of the queue and evict it.
+        if (recordsToCheck != null && recordsToCheck.batchUniqueIdentifier()
+                .equals(recordsDeliveryAck.batchUniqueIdentifier())) {
+            evictNextResult();
+            // Upon evicting, check if queue is empty. if yes, then give the drain control back to publisher thread.
+            if(getRecordsResultQueue.isEmpty()) {
+                log.debug("{} : Publisher thread takes over the draining control. Queue Size : {}, Demand : {}", shardId,
+                        getRecordsResultQueue.size(), requestedResponses.get());
+                shouldDrainEventOnAck.set(false);
+            } else {
+                // Else attempt to drain the queue.
+                drainQueueForRequests();
```

Comment:
In the happy case where we continue to receive acks for each event delivered, always the ack-notifier thread will be draining the queue. Publisher thread will continue to do only the fetch part. I think this conflicts a bit with the design of this class and the purpose of the publisher thread.

Reply:
Publisher thread will be a kick starter in the event of paused deliveries. In happy case, yes ack-notifier will be scheduling for minimum delay between each delivery and for simplicity.

Reply:
I agree that it'll help a bit with the delay. But not with the simplicity. :) We have 3 different threads draining the queue. I think it would be better and simpler if we had only one thread publishing and other two just updating the demand and registering the ack. Anyway that's a bigger change and we can consider that in the future. :)

Reply:
Addressed this in https://github.com/awslabs/amazon-kinesis-client/pull/606/files